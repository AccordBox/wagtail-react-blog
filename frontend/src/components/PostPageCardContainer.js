import React from "react";
import { PostPageCard } from "./PostPageCard";
import { classNames } from "../utils";
import { useLocation, Link } from "react-router-dom";

function getPageItems(props, location) {
  const { paginator } = props;
  const { currentPage, numPages } = paginator;
  let items = [];

  const curPath = location.pathname;

  let prePageUrl, nextPageUrl;
  if (curPath.match(/\/page-[0-9]+/)) {
    prePageUrl = curPath.replace(/\/page-[0-9]+/, `/page-${currentPage - 1}`);
    nextPageUrl = curPath.replace(/\/page-[0-9]+/, `/page-${currentPage + 1}`);
  } else {
    prePageUrl = `${curPath}/page-${currentPage - 1}`.replace("//", "/");
    nextPageUrl = `${curPath}/page-${currentPage + 1}`.replace("//", "/");
  }

  items.push(
    <li key={prePageUrl}>
      <Link
        to={prePageUrl}
        className={classNames(
          currentPage <= 1
            ? "pointer-events-none text-gray-300"
            : "text-blue-500 dark:text-white",
          "inline-block py-2 px-4 bg-white  border border-gray-300 ",
          "border-r-0 rounded-l-lg hover:bg-blue-500 hover:text-white ",
          "dark:bg-gray-700 dark:border-gray-500 dark:hover:bg-blue-500 dark:hover:text-white"
        )}
      >
        Prev
      </Link>
    </li>
  );

  items.push(
    <li key={nextPageUrl}>
      <Link
        to={nextPageUrl}
        className={classNames(
          currentPage >= numPages
            ? "pointer-events-none text-gray-300"
            : "text-blue-500 dark:text-white ",
          "inline-block py-2 px-4 bg-white  border border-gray-300 ",
          "rounded-r-lg hover:bg-blue-500 hover:text-white ",
          "dark:bg-gray-700 dark:border-gray-500 dark:hover:bg-blue-500 dark:hover:text-white"
        )}
      >
        Next
      </Link>
    </li>
  );

  return items;
}

function getFilterMsg(props) {
  const { filterMeta } = props;

  let filterMsg = "";
  if (filterMeta.filterType) {
    filterMsg = (
      <div
        className="px-4 py-3 leading-normal text-blue-700 bg-blue-100 rounded-lg mb-4"
        role="alert"
      >
        <p>
          {filterMeta.filterType}: {filterMeta.filterTerm}
        </p>
      </div>
    );
  }
  return filterMsg;
}

function PostPageCardContainer(props) {
  const { childrenPages } = props;
  const location = useLocation();
  const pageItems = getPageItems(props, location);
  const filterMsg = getFilterMsg(props);

  return (
    <main role="main" className="w-full sm:w-2/3 md:w-3/4 lg:w-8/12 px-2 mb-4">
      {filterMsg}

      {childrenPages.map((post) => (
        <PostPageCard post={post} key={post.pageContent.id} />
      ))}

      <nav aria-label="Page navigation">
        <ul className="flex pl-0 rounded list-none flex-wrap">{pageItems}</ul>
      </nav>
    </main>
  );
}

export { PostPageCardContainer };
