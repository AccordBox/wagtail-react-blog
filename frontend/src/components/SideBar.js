import React from "react";
import { TagWidget } from "./TagWidget";
import { CategoryWidget } from "./CategoryWidget";
import { SearchForm } from "./SearchForm";
import { BookCard } from "./BookCard";

function SideBar(props) {
  return (
    <div className="w-full sm:w-1/3 md:w-1/4 lg:w-4/12 px-2">
      <SearchForm {...props} />
      <CategoryWidget {...props} />
      <TagWidget {...props} />
      <BookCard/>
    </div>
  );
}

export { SideBar };
