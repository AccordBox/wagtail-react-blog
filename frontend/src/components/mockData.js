export const blogPageData = {
  page_type: "BlogPage",
  page_content: {
    id: 3,
    slug: "blogpage",
    title: "BlogPage",
    url: "/",
    last_published_at: "2021-08-11T02:44:45.918000Z"
  },
  categories_list: [
    {
      name: "Programming",
      slug: "programming",
      url: "/category/programming/"
    },
    {
      name: "Life",
      slug: "life",
      url: "/category/life/"
    }
  ],
  tags_list: [
    {
      name: "Django",
      slug: "django",
      url: "/tag/django/"
    },
    {
      name: "Wagtail",
      slug: "wagtail",
      url: "/tag/wagtail/"
    },
    {
      name: "React",
      slug: "react",
      url: "/tag/react/"
    }
  ],
  children_pages: [
    {
      page_type: "PostPage",
      page_content: {
        id: 4,
        slug: "postpage1",
        title: "PostPage1",
        url: "/postpage1/",
        last_published_at: "2021-08-11T02:52:16.766000Z",
        tags: [
          {
            name: "Django",
            slug: "django",
            id: 1
          }
        ],
        categories: [
          {
            name: "Programming",
            slug: "programming",
            id: 1
          }
        ],
        body: [
          {
            type: "h1",
            value: "The Zen of Wagtail",
            id: "886b6fb8-472a-47b0-9bef-2317ea04799a"
          },
          {
            type: "paragraph",
            value: "<p>Wagtail has been born out of many years of experience building websites, learning approaches that work and ones that don’t, and striking a balance between power and simplicity, structure and flexibility. We hope you’ll find that Wagtail is in that sweet spot.</p>",
            id: "6b051ad8-d1ba-4dab-84be-1d30b4e272e9"
          },
          {
            type: "image_carousel",
            value: [
              {
                url: "/media/images/image_2.original.jpg",
                width: 1280,
                height: 853,
                alt: "image_2.jpeg"
              },
              {
                url: "/media/images/image_1.original.jpg",
                width: 1280,
                height: 853,
                alt: "image_1.jpeg"
              }
            ],
            id: "7231f723-5770-4173-a17e-a86a0b68bac1"
          },
          {
            type: "image_text",
            value: {
              reverse: false,
              text: "<p>Wagtail is not an instant website in a box.</p><p>You can’t make a beautiful website by plugging off-the-shelf modules together - expect to write code.</p>",
              image: {
                url: "/media/images/image_3.width-800.jpg",
                width: 800,
                height: 533,
                alt: "image_3.jpeg"
              }
            },
            id: "b32bc7bb-0e71-448b-968d-67be3492659f"
          },
          {
            type: "image_text",
            value: {
              reverse: true,
              text: "<p>A CMS should get information out of an editor’s head and into a database, as efficiently and directly as possible.</p>",
              image: {
                url: "/media/images/image_1.width-800.jpg",
                width: 800,
                height: 533,
                alt: "image_1.jpeg"
              }
            },
            id: "ab6f3090-b05d-4c07-a3b0-c105ddf823a6"
          }
        ],
        header_image: {
          url: "/media/images/image_3.max-1000x800.jpg",
          width: 1000,
          height: 666,
          alt: "image_3.jpeg"
        },
        content_type_str: "blog.postpage"
      }
    },
    {
      page_type: "PostPage",
      page_content: {
        id: 5,
        slug: "postpage2",
        title: "PostPage2",
        url: "/postpage2/",
        last_published_at: "2021-08-11T02:51:51.605000Z",
        tags: [
          {
            name: "React",
            slug: "react",
            id: 3
          }
        ],
        categories: [],
        body: [
          {
            type: "h1",
            value: "The Zen of Wagtail",
            id: "886b6fb8-472a-47b0-9bef-2317ea04799a"
          },
          {
            type: "paragraph",
            value: "<p>Wagtail has been born out of many years of experience building websites, learning approaches that work and ones that don’t, and striking a balance between power and simplicity, structure and flexibility. We hope you’ll find that Wagtail is in that sweet spot.</p>",
            id: "6b051ad8-d1ba-4dab-84be-1d30b4e272e9"
          },
          {
            type: "image_carousel",
            value: [
              {
                url: "/media/images/image_2.original.jpg",
                width: 1280,
                height: 853,
                alt: "image_2.jpeg"
              },
              {
                url: "/media/images/image_1.original.jpg",
                width: 1280,
                height: 853,
                alt: "image_1.jpeg"
              }
            ],
            id: "7231f723-5770-4173-a17e-a86a0b68bac1"
          },
          {
            type: "image_text",
            value: {
              reverse: false,
              text: "<p>Wagtail is not an instant website in a box.</p><p>You can’t make a beautiful website by plugging off-the-shelf modules together - expect to write code.</p>",
              image: {
                url: "/media/images/image_3.width-800.jpg",
                width: 800,
                height: 533,
                alt: "image_3.jpeg"
              }
            },
            id: "b32bc7bb-0e71-448b-968d-67be3492659f"
          },
          {
            type: "image_text",
            value: {
              reverse: true,
              text: "<p>A CMS should get information out of an editor’s head and into a database, as efficiently and directly as possible.</p>",
              image: {
                url: "/media/images/image_1.width-800.jpg",
                width: 800,
                height: 533,
                alt: "image_1.jpeg"
              }
            },
            id: "ab6f3090-b05d-4c07-a3b0-c105ddf823a6"
          }
        ],
        header_image: {
          url: "/media/images/image_2.max-1000x800.jpg",
          width: 1000,
          height: 666,
          alt: "image_2.jpeg"
        },
        content_type_str: "blog.postpage"
      }
    }
  ],
  paginator: {
    per_page: 2,
    current_page: 1,
    num_pages: 2
  },
  filter_meta: {
    filter_type: null,
    filter_term: null
  }
}
