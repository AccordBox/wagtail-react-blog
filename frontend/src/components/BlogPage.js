import React from "react";
import { Container, Row } from "react-bootstrap";
import { TopNav } from "./TopNav";
import { Footer } from "./Footer";
import { PostPageCardContainer } from "./PostPageCardContainer";
import { SideBar } from "./SideBar";

class BlogPage extends React.Component {
  render() {
    return (
      <div>
        <TopNav />
        <Container>
          <Row>
            <PostPageCardContainer {...this.props} />
            <SideBar />
          </Row>
        </Container>
        <Footer />
      </div>
    );
  }
}

export { BlogPage };
