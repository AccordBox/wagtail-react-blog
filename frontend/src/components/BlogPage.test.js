import React from "react";
import BlogPage from "./BlogPage";
import { screen, render } from "@testing-library/react";
import { blogPageData } from "./mockData";
import { PostPageCardContainer } from "./PostPageCardContainer";
import { SideBar } from "./SideBar";
import camelcaseKeys from "camelcase-keys";
import { MemoryRouter } from "react-router-dom";

jest.mock("./PostPageCardContainer", () => ({
  PostPageCardContainer: jest.fn(({ children }) => (
    <div data-testid="PostPageCardContainer">{children}</div>
  )),
}));

jest.mock("./SideBar", () => ({
  SideBar: jest.fn(({ children }) => (
    <div data-testid="SideBar">{children}</div>
  )),
}));

test("BlogPage Test", () => {
  const data = camelcaseKeys(blogPageData, { deep: true });

  const { asFragment } = render(
    <MemoryRouter>
      <BlogPage {...data} />
    </MemoryRouter>
  );

  expect(screen.queryByTestId("PostPageCardContainer")).toBeInTheDocument();
  expect(PostPageCardContainer).toHaveBeenCalledWith(data, expect.anything());

  expect(screen.queryByTestId("SideBar")).toBeInTheDocument();
  expect(SideBar).toHaveBeenCalledWith(data, expect.anything());

  expect(asFragment()).toMatchSnapshot();
});
