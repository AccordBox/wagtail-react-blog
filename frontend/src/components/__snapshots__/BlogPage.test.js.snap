// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPage Test 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col min-h-screen"
  >
    <nav
      class="bg-white border-b border-opacity-75 border-gray-300 dark:bg-gray-900 dark:text-white dark:border-gray-500"
    >
      <div
        class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8"
      >
        <div
          class="relative flex items-center justify-between h-16"
        >
          <div
            class="absolute inset-y-0 left-0 flex items-center sm:hidden"
          >
            <button
              aria-expanded="false"
              class="inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-white"
              id="headlessui-disclosure-button-1"
              type="button"
            >
              <span
                class="sr-only"
              >
                Open main menu
              </span>
              <svg
                aria-hidden="true"
                class="block h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6h16M4 12h16M4 18h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start"
          >
            <div
              class="flex-shrink-0 flex items-center"
            >
              <a
                href="/"
              >
                <span
                  class="text-bold text-grey-800"
                >
                  React Wagtail Demo
                </span>
              </a>
            </div>
            <div
              class="hidden sm:block sm:ml-6"
            >
              <div
                class="flex space-x-4"
              >
                <a
                  class="text-gray-500 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium"
                  href="#"
                >
                  Dashboard
                </a>
                <a
                  class="text-gray-500 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium"
                  href="#"
                >
                  Team
                </a>
                <a
                  class="text-gray-500 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium"
                  href="#"
                >
                  Projects
                </a>
                <a
                  class="text-gray-500 hover:bg-gray-700 hover:text-white px-3 py-2 rounded-md text-sm font-medium"
                  href="#"
                >
                  Calendar
                </a>
              </div>
            </div>
          </div>
          <div
            class="absolute inset-y-0 right-0 flex items-center pr-2 sm:static sm:inset-auto sm:ml-6 sm:pr-0"
          >
            <button
              class="bg-gray-800 p-1 rounded-full text-gray-400 hover:text-white focus:outline-none"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003 0 0012 21a9.003 9.003 0 008.354-5.646z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="w-full max-w-7xl mx-auto px-2 sm:px-6 lg:px-8"
    >
      <div
        class="flex flex-row flex-wrap py-4"
      >
        <div
          data-testid="PostPageCardContainer"
        />
        <div
          data-testid="SideBar"
        />
      </div>
    </div>
    <footer
      class="mt-auto py-7 bg-white border-t border-opacity-75 border-gray-300 dark:bg-gray-900 dark:text-white dark:border-gray-500"
    >
      <div
        class="container mx-auto"
      >
        <p
          class="m-0 text-center text-blue-700 hover:underline dark:text-gray-400"
        >
          Built by 
          <a
            href="https://www.accordbox.com/"
          >
            michaelyin@accordbox.com
          </a>
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
